### PR TITLE
Rewrote the grad_sample tests to use Hypothesis

### PR DESCRIPTION
**Pull request by touqir14**
This PR addresses the [issue](https://github.com/pytorch/opacus/issues/122) where it was requested to migrate the Grad Sample tests in the directory ```opacus/tests/grad_samples``` to using hypothesis in order to make it much easier to write new and extend the current test cases. The test configuration space of the added code is at least as large (larger in most cases) as in the old testing code. I can push another commit with this PR to add grad sample test cases for PackedSequence usage in the DPLSTM class once the PackedSequence [functionalities](https://github.com/pytorch/opacus/pull/116) are merged.

### CODE REVIEW

File: opacus/tests/grad_samples/conv1d_test.py [hunk 1]
```diff
@@ -1,93 +1,56 @@
...
-    def test_3d_input_shrinking_12_groups_kernel2(self):
-        N, C, W = 32, 24, 10
-        x = torch.randn([N, C, W])
-        conv = nn.Conv1d(C, C // 2, kernel_size=2, groups=12)
+class Conv1d_test(GradSampleHooks_test):
+    @given(
+        N=st.sampled_from([32]),
+        C=st.sampled_from([3, 4, 12, 24]),
+        W=st.sampled_from([11, 10]),
+        out_channels_mapper=st.sampled_from([expander, shrinker]),
```
Comment (Darktex) [inline opacus/tests/grad_samples/conv1d_test.py:22]: Smart!

File: opacus/tests/grad_samples/embedding_test.py [hunk 2]
```diff
@@ -4,38 +4,45 @@
...
+        N: int,
+        T: int,
+        Q: int,
+        R: int,
+        V: int,
+        D: int,
+        dim: int,
+    ):
+
+        if dim == 1:
```
Comment (Darktex) [inline opacus/tests/grad_samples/embedding_test.py:37]: Love it

File: opacus/tests/grad_samples/group_norm_test.py [hunk 3]
```diff
@@ -4,29 +4,35 @@
...
+from typing import Optional, Tuple, Callable, Union
+
+import hypothesis.strategies as st
+from hypothesis import given, settings
+
 from .common import GradSampleHooks_test
 
 
 class GroupNorm_test(GradSampleHooks_test):
-    """
```
Comment (Darktex) [inline opacus/tests/grad_samples/group_norm_test.py:11]: nit: Keep this comment?

File: opacus/tests/grad_samples/group_norm_test.py [hunk 4]
```diff
@@ -4,29 +4,35 @@
...
-        x = torch.randn([N, C, H, W])
-        norm = nn.GroupNorm(num_groups=1, num_channels=C, affine=True)
-        self.run_test(x, norm, batch_first=True)
-
-    def test_3d_input_four_groups_affine(self):
-        N, C, H, W = 32, 16, 8, 10
-        x = torch.randn([N, C, H, W])
-        norm = nn.GroupNorm(num_groups=4, num_channels=C, affine=True)
-        self.run_test(x, norm, batch_first=True)
+        if num_groups == "C":
```
Comment (Darktex) [inline opacus/tests/grad_samples/group_norm_test.py:33]: (y)

**Review by Darktex — CHANGES_REQUESTED** [review]
You are on fire @touqir14! Yet another awesome diff :) 

With this diff you translated our existing tests and we now test exactly the same things with Hypothesis. While this is already awesome as we gain A TON of readability, one of the advantages of Hypothesis is also in testing *more* stuff and looking for edge cases that you may not normally think (a typical one is, of course, the pesky batch_size=1).

Using `st.sampled()` defeats this purpose: it's actually better to use integers in a range and let Hypothesis do its own thing. Check out how we test DPLSTM with Hypothesis: https://github.com/pytorch/opacus/blob/master/opacus/tests/dp_layers/dp_lstm_test.py#L30

File: opacus/tests/grad_samples/group_norm_test.py (hunk 3 above)
Comment (touqir14) [reply opacus/tests/grad_samples/group_norm_test.py:11]: Right. Since I created this from a new file, I must have forgotten to copy and paste it

**Comment by touqir14** [comment]
> You are on fire @touqir14! Yet another awesome diff :)
> 
> With this diff you translated our existing tests and we now test exactly the same things with Hypothesis. While this is already awesome as we gain A TON of readability, one of the advantages of Hypothesis is also in testing _more_ stuff and looking for edge cases that you may not normally think (a typical one is, of course, the pesky batch_size=1).
> 
> Using `st.sampled()` defeats this purpose: it's actually better to use integers in a range and let Hypothesis do its own thing. Check out how we test DPLSTM with Hypothesis: https://github.com/pytorch/opacus/blob/master/opacus/tests/dp_layers/dp_lstm_test.py#L30

Using the range strategies are definitely more thorough. I originally opted for ```st.sampled()``` as I feared widening the configuration space can slow down the tests significantly.

**Comment by Darktex** [comment]
It's fine :) We mark a timeout for that, and Hypothesis will do its best to explore the space within that budget

**Comment by touqir14** [comment]
@Darktex , So I have changed some of the ```st.sampled_from``` to ```st.integers```. Now ```conv1d_test.py``` and ```conv2d_test.py``` are failing. The error from ```conv1d_test.py``` is ```RuntimeError: NNPACK SpatialConvolution_updateOutput failed``` which is due to an NNPACK bug, related to Pytorch, exposed [here](https://discuss.pytorch.org/t/cant-run-nn-conv1d-with-libtorch-on-ios/71719) . The errors from ```conv2d_test.py``` is probably due to an opacus bug :
```
AssertionError: A total of 2 values do not match for loss_reduction=mean: 
	1. ('Param weight: Opacus L2 norm = : 1.0874651670455933, ', 'Microbatch L2 norm = : 1.0874567031860352, ', 'MSE = 3.651851043784404e-11, ', 'L1 Loss = 1.2964130746695446e-07')
	2. ('Param bias: Opacus L2 norm = : 0.0770709440112114, ', 'Microbatch L2 norm = : 0.07706724852323532, ', 'MSE = 3.668529716116531e-11, ', 'L1 Loss = 1.617791554053838e-07')
```
which is caused due to the inputs:
```
    N=32,
    C=3,
    H=9,
    W=9,
    out_channels_mapper=<function opacus.tests.grad_samples.conv2d_test.<lambda>>,
    kernel_size=2,
    stride=1,
    padding=2,
    groups=1,
```
We also see ```RuntimeError: NNPACK SpatialConvolution_updateOutput failed``` error in this test file too.
